### ISSUE TITLE
Feature: online_update aka you

### ISSUE THREAD

**Issue by manfredsc** [shundhammer/myrlyn#2]
Hi,
thanks for your project, I appreciate it!
I played a bit with it, it seems UI-wise the project is almost complete and works fine.
Great job!

The online_update tool has almost the same interface as sw_single, so I wondered
if the "you" part of yast could also be supported (I.e. the "patches" tab). I find it very useful.
It is an additional benefit over zypper in the "what-if" sense, I often modify the suggested package list,
de-select some additionally recommended packages, read the patch changelog and decide
if I already want to update a specific package. With "you" all information is present at a glance,
in one window.

It would be great if "you" could be added to this project. Thanks!

**Comment by shundhammer** [comment]
It's already there (and it has been in yast-qt sw_single all the time): Hit the F2 key, and voila, you have the "Patches" view. Try it!

In yqpkg, it still has some weird issues that I haven't had time yet to follow through. But in principle, it's there; and once packages (well, really "resolvables" in libzypp speak) can be installed, the patches should simply be there.

There is also a "simple patch selector" in the attic/ subdir; that's the real thing that you'd get for `yast online_update`. But that's actually just a stripped down package selector where that patches view is there from the start, and it's the main view.

HTH

**Comment by shundhammer** [comment]
Having said that, my original plan was to enable the "Patches" view by default _if there are any patches_ in the libzypp pool; there is a `#define` for that. But right now, that dumps a gazillion messages into the log, which is really annoying for development.

There must be some probelm with the `YQPkgSelMapper`; that's a class to map from one libzypp type to another because that one doesn't have a _selectable_ as a bracket object holding together several _resolvables_ (typically packages; one specific version from one repo for each). I don't recall the details anymore; that will need some research.

I wrote that thing a long time ago; but that must be about 20 years in the past. Yikes.

**Comment by manfredsc** [comment]
Wow, what an Easter Egg! I use yast for more than 20 years now (probably rather 25 years), and it never occurred to me,
that you can get the "patches" tab to appear in sw_single.

Well, I know why it was never advertised: It does not work. I tried it on Leap 15.6, with the real yast.
It works for the first time, but when you invoke sw_single a second time, you get an orange pop-up with 
"UI Syntax Error Out of Memory", and in the log file there is
```
2024-12-06 01:02:28 <2> localhost.localdomain(6172) [qt-pkg] YQPkgFilterTab.cc(loadSettings):670 No page with ID "patches"
2024-12-06 01:02:28 <2> localhost.localdomain(6172) [ui] YQPkgFilterTab.cc(showPage):326        THROW:    Null pointer
2024-12-06 01:02:28 <1> localhost.localdomain(6172) [qt-pkg] YQPackageSelectorBase.cc(~YQPackageSelectorBase):102 Destroying PackageSelector
2024-12-06 01:02:28 <3> localhost.localdomain(6172) [qt-pkg] YQPackageSelectorPluginImpl.cc(createPackageSelector):61 Caught std::exception: Null pointer
2024-12-06 01:02:28 <3> localhost.localdomain(6172) [qt-pkg] YQPackageSelectorPluginImpl.cc(createPackageSelector):62 This is a libzypp problem. Do not file a bug against the UI!
2024-12-06 01:02:28 <2> localhost.localdomain(6172) [ui] YQPackageSelectorPluginImpl.cc(createPackageSelector):70       THROW:    Out of memory
2024-12-06 01:02:28 <2> localhost.localdomain(6172) [ui] YCP_UI.cc(OpenDialog):584      CAUGHT:   Out of memory
2024-12-06 01:02:28 <3> localhost.localdomain(6172) [libycp] modules/PackagesUI.rb:312 UI::OpenDialog() failed
2024-12-06 01:02:28 <1> localhost.localdomain(6172) [ui] YPushButton.cc(setFunctionKey):202 Guessing button role YOKButton for YPushButton "Close" at 0x7fe1f43f5480 from function key F10
```

Luckily, on the third invocation, sw_single works again, but without the "patches" tab.

**Comment by shundhammer** [comment]
I tried this yesterday in yqpkg with adding the patches view with the `F2' key, and I could select the 3 pending patches for my Leap 15.6 and install them without a problem.

Not sure what happened in your case, but it appears to work in principle.

**Comment by manfredsc** [comment]
As said, it works for me, too.
It crashes in the second round for me, after closing yast2 and restarting it again.
Then, it wants to read the tab configuration from disk, I guess, and has
trouble adding the "patches" tab to the UI, and shouts with the above error and backtrace.
The tab configuration is remembered in yast2, and it opens the same
tabs and in the same order as in the last invocation.
And the crash seems to clean up after itself by deleting the "patches" entry
from the configuration stored on disk. So in the third invocation of yast2,
everything is fine again.

**Comment by manfredsc** [comment]
To clarify, this was with the *real* yast2 on Leap 15.6.
I just tried it with the current checkout of yqpkg, there is no visible crash, a new invocation just
starts without a "patches" tab.
But in the log, there is the same issue as with yast2, I get the following log entries with yqpkg:

```
[...]
2024-12-20 11:19:57.344 [38460] <Debug>   YQPkgServiceList.cc:70 YQPkgServiceList():  Creating service list done
2024-12-20 11:19:58.218 [38460] <Debug>   YQPkgFilterTab.cc:657 readSettings():  Restoring pages search, repos, patterns, package_classification, patches, inst_summary
2024-12-20 11:19:58.218 [38460] <Debug>   YQPkgFilterTab.cc:658 readSettings():  Current page: patches
2024-12-20 11:19:58.220 [38460] <WARNING> YQPkgFilterTab.cc:667 readSettings():  No page with ID "patches"
2024-12-20 11:19:58.220 [38460] <WARNING> YQPkgFilterTab.cc:677 readSettings():  Can't restore current page with ID "patches"
2024-12-20 11:19:58.227 [38460] <Debug>   YQPkgSelector.cc:254 YQPkgSelector():  YQPkgSelector init done
2024-12-20 11:19:58.227 [38460] <Debug>   MainWindow.cc:89 showPage():  Showing class YQPkgSelector
[...]
```

**Comment by shundhammer** [comment]
Yes, re-initializing all of the package selector after a package commit and then going back is something that I am currently working on. It's getting better by the hour. ;-)

I wouldn't be surprised if the patches list would need some more drastic re-init; a stronger kick in the butt. ;-) But the whole patches thing will have to wait until the main features work smoothly. But it's on the radar.

**Comment by shundhammer** [comment]
BTW since you bravely tried _yqpkg_, did you try the "Details" view on the commit page? Was that more helpful or more confusing to you?

And how about the current summary page? Does it contain the information that you'd be looking for, or too much, or too little?

I considered for a short moment if those things should contain the whole she-bang with package name + version and maybe even arch, and then I decided against it; personally, I am never interested in the exact version number, that only distracts me while I am trying to make sense of what's going on: "The kernel is being downloaded", "the kernel is being installed" and later "I installed a new kernel" is just enough information for me. How about you?

**Comment by manfredsc** [comment]
Details: well, if there are only a few packages to install, things go so fast that installation is finished before you can 
realize and track all the windows. But if you have to install a lot of packages, this detailed view is actually nice, IMO.
One could even make the detailed view the default choice, IMO.

Summary page: If you have activated the detailed view, the summary page really serves for nothing, I would prefer to
stay on the details page then, with the added countdown. I.e. just display the additional stop button etc below the detailed view, after the installation finished.

Display of package names: names-only is mostly enough, but when you have multiversion packages, this can be confusing.
I.e. if you install two kernels, you get twice "kernel-default". So if I could choose, I would display full name+version for multiversion packages and name-only for the rest. But I realize that this could be an additional, not really needed complication.

PS: In the "patches" tab, all package entries appear twice for me...

**Comment by shundhammer** [comment]
## Patches View Available

If the system provides any patches, the "Patches" view is now automatically available:

![Myrlyn-patches-view](https://github.com/user-attachments/assets/4f499cd3-4d1f-4007-b7c4-06d369850633)



## Number of Patches / Updates in the Tab Labels

You can still use the "Updates" view if you prefer that one. Both "Patches" and "Updates" now show the number of available patches (or package updates, respectively) in their tab labels so you can instantly see if it's even worthwhile to click into them:

![Myrlyn-updates-view](https://github.com/user-attachments/assets/a7c8d54b-e374-4a70-b196-9daa9691aac8)

Notice that not all repos offer patches, even for distros that have them, like Leap; especially third-party repos like Packman or NVidia simply provide updated packages.

But even for repos that do offer patches, you can also simply update all the packages that belong to them in the "Updates" view: A patch is _satisfied_ if all the package versions that it provdes are installed.
